map/cloud.cpp

```diff
@@ -0,0 +1,485 @@
+}
+
+void Cloud::SetSynchronizationHandlers(SynchronizationStartedHandler && onSynchronizationStarted,
+                                       SynchronizationFinishedHandler && onSynchronizationFinished)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  m_onSynchronizationStarted = std::move(onSynchronizationStarted);
+  m_onSynchronizationFinished = std::move(onSynchronizationFinished);
+}
+
+void Cloud::SetState(State state)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  m_state = state;
+  settings::Set(m_settingsParamName, static_cast<int>(m_state));
+
+  switch (m_state)
+  {
+    case State::Enabled:
+      GetPlatform().RunTask(Platform::Thread::File, [this]() { LoadIndex(); });
+      break;
+
+    case State::Disabled:
+      // Delete index file and clear memory.
+      my::DeleteFileX(m_indexName);
+      m_files.clear();
+      m_index = Index();
+      break;
+
+    case State::Unknown:
+      ASSERT(false, ("Unknown state can't be set up"));
+  }
+}
+
+Cloud::State Cloud::GetState() const
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  return m_state;
+}
+
+void Cloud::Init(std::vector<std::string> const & filePaths)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  for (auto const & path : filePaths)
+    m_files.insert(path);
+
+  if (m_state != State::Enabled)
+    return;
+
+  GetPlatform().RunTask(Platform::Thread::File, [this]() { LoadIndex(); });
+}
+
+void Cloud::MarkModified(std::string const & filePath)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  if (m_state != State::Enabled)
+    return;
+
+  m_files.insert(filePath);
+  MarkModifiedImpl(filePath, false /* checkSize */);
+}
+
+std::unique_ptr<User::Subscriber> Cloud::GetUserSubscriber()
+{
+  auto s = std::make_unique<User::Subscriber>();
+  s->m_onChangeToken = [this](std::string const & token)
+  {
+    SetAccessToken(token);
+    ScheduleUploading();
+    return User::Subscriber::Action::DoNothing;
+  };
+  return s;
+}
+
+void Cloud::LoadIndex()
+{
+  auto const indexFilePath = GetIndexFilePath(m_indexName);
+  if (GetPlatform().IsFileExistsByFullPath(indexFilePath))
+  {
+    // Read index file.
+    std::string data;
+    try
+    {
+      FileReader r(indexFilePath);
+      r.ReadAsString(data);
+    }
+    catch (FileReader::Exception const & exception)
+    {
+      data.clear();
+      LOG(LWARNING, ("Exception while reading file:", indexFilePath,
+                     "reason:", exception.what()));
+    }
+
+    // Parse index file.
+    if (!data.empty())
+    {
+      try
+      {
+        Index index;
+        coding::DeserializerJson deserializer(data);
+        deserializer(index);
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        std::swap(m_index, index);
+      }
+      catch (my::Json::Exception const & exception)
+      {
+        LOG(LWARNING, ("Exception while parsing file:", indexFilePath,
+          "reason:", exception.what()));
+      }
+    }
+  }
+
+  UpdateIndex();
```

Comment:
UpdateIndex and SaveIndex both enters critical section, is it possible to enter critical section once?

Reply:
+1

Reply:
done

Reply:
optional: ScheduleUploading also enters critical section. But not sure that we need to combine this with UpdateIndex.